pkg/sql/logictest/testdata/logic_test/udf_params

```diff
@@ -593,3 +593,147 @@ SELECT b FROM f_udt()
+
+statement ok
+CREATE FUNCTION my_sum(a INT, b INT, c INT) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b + c; $$;
+
+statement error pgcode 42P13 input parameters after one with a default value must also have defaults
+CREATE OR REPLACE FUNCTION my_sum(a INT = 1, b INT, c INT = 3) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b + c; $$;
+
+# Adding a default expression to a parameter is ok.
+statement ok
+CREATE OR REPLACE FUNCTION my_sum(a INT, b INT DEFAULT 2, c INT = 3) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b + c; $$;
+
+# But removing an existing default expression is not ok.
+statement error pgcode 42P13 cannot remove parameter defaults from existing function
+CREATE OR REPLACE FUNCTION my_sum(a INT, b INT, c INT = 3) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b + c; $$;
+
+statement error pgcode 22P02 could not parse "a" as type int
+CREATE OR REPLACE FUNCTION my_sum(a INT = 'a', b INT DEFAULT 2, c INT = 3) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b + c; $$;
+
+query I
+SELECT my_sum(1);
+----
+6
+
+query I
+SELECT my_sum(1, 1);
+----
+5
+
+query I
+SELECT my_sum(1, 1, 1);
+----
+3
+
+# Note that postgres returns "function my_sum() does not exist".
+statement error pgcode 42883 unknown signature: public.my_sum\(\)
+SELECT my_sum();
+
+# Note that postgres returns "function my_sum(int, int, int, int) does not exist".
+statement error pgcode 42883 unknown signature: public.my_sum\(int, int, int, int\)
+SELECT my_sum(1, 1, 1, 1);
+
+# Same as above, but the default value needs to be coerced from numeric to int
+# (becoming 4).
+statement ok
+CREATE OR REPLACE FUNCTION my_sum(a INT, b INT DEFAULT 2, c INT = 3.5) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b + c; $$;
+
+query I
+SELECT my_sum(1);
+----
+7
+
+query I
+SELECT my_sum(1, 1);
+----
+6
+
+# Add another overload that creates ambiguity for some number of input
+# arguments.
+statement ok
+CREATE FUNCTION my_sum(a INT) RETURNS INT LANGUAGE SQL AS $$ SELECT a; $$;
+
+statement error pgcode 42725 function name "my_sum" is not unique
+SELECT my_sum(1);
+
+statement ok
+DROP FUNCTION my_sum(INT);
+
+statement ok
+DROP FUNCTION my_sum;
+
+# Create a new function with the same name as the one we just dropped and then
+# invoke in the same way as above - the optimizer should realize that the cached
+# memo is stale.
+statement ok
+CREATE FUNCTION my_sum(OUT sum INT, INOUT a INT, INOUT b INT = 3) AS $$ SELECT (a + b, a, b); $$ LANGUAGE SQL;
+
+query T
+SELECT my_sum(1);
+----
+(4,1,3)
+
+query III
+SELECT * FROM my_sum(1);
+----
+4  1  3
+
+query T
+SELECT my_sum(1, 1);
+----
+(2,1,1)
+
+query III
+SELECT * FROM my_sum(1, 1);
+----
+2  1  1
+
+statement ok
+DROP FUNCTION my_sum;
+
+# Test case when DEFAULT expression uses a UDF.
+
+statement ok
+CREATE FUNCTION f1(a INT, b INT = 2) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b; $$;
+
+statement ok
+CREATE FUNCTION f2(a INT, b INT = f1(1)) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b; $$;
```

Comment:
can you also add tests for having a default for an enum type and composite type?

that should also be tested in the random schema change workload; not sure if you want to do that now, or file a ticket for later.